### ISSUE TITLE
Android application scope getting reset after test

### ISSUE THREAD

**Issue by efung** [stephanenicolas/toothpick#188]
Suppose I create a global application scope from my Android `Application` class, e.g.

```
    @Override
    public void onCreate() {
        super.onCreate();

        Scope appScope = Toothpick.openScopes(APP_SCOPE);
        appScope.installModules(
                new SmoothieApplicationModule(this),
                new MyAppModule(this));
        );
    }
```

Suppose this global scope is used everywhere in my app.

In my instrumentation tests, I want to mock out some of the instances provided by `MyAppModule`. If I use `ToothPickRule` in my tests, so that I get the auto-binding of `@Mock`s from `ToothPickTestModule`, the entire Toothpick scope gets reset after each test. Since the `Application.onCreate()` is only called once, I lose any modules I installed in that scope.

What would be the recommended approach here?

**Comment by efung** [comment]
Instead of `reset()`, would it make sense to call `closeScope`  on the scope name set on `ToothPickRule`?

**Comment by efung** [comment]
> Instead of `reset()`, would it make sense to call `closeScope`  on the scope name set on `ToothPickRule`?

Actually, that won't work. My thinking was that I could call `openScopes` with a test scope being a child of the (global) application scope, then install the test modules on the test scope. However, the rest of the application wouldn't be able to see the test modules.

**Comment by dlemures** [comment]
Hi @efung ,

The rule was developed with local unit testing in mind.

Just thinking out loud, something that you could do is:
1. Using a child scope in production:
```java
@Override
public void onCreate() {
    super.onCreate();

    Scope rootScope = Toothpick.openScopes(ROOT_SCOPE);
    rootScope.installModules(
            new SmoothieApplicationModule(this),
            new MyAppModule(this));
    );
    Scope appScope = Toothpick.openScopes(ROOT_SCOPE, APP_SCOPE); // Used for injection throughout the app
}
```
2. And using `ToothPickTestModule` directly in test:
```java
@Before 
public void setUp() {
    Scope appScope = Toothpick.openScopes(ROOT_SCOPE, APP_SCOPE);
    appScope.installTestModules(new ToothPickTestModule(this));
}

@Before 
public void tearDown() {
    Toothpick.closeScope(APP_SCOPE);
}
```

You can do step 1 just for `androidTest`, by wrapping the logic to open the injection scope.

**Comment by dlemures** [comment]
@efung did it work for you?
I'm wondering if we should add information about this to the wiki.

**Comment by efung** [comment]
@dlemures Yes, that approach of opening a child scope is working very well for us in our instrumentation tests.

**Comment by kanawish** [comment]
A small note/gotcha for visitors to this issue. Having a child scope implies you might need a custom `@Singleton` annotation. By default, `@Singleton` annotated objects are automatically attached to the root scope here.

**Comment by stephanenicolas** [comment]
@Singleton is THE annotation that allows to create Singletons for the root
scope. Any other scope needs a custom annotation and to bind the annotation
scope to the scope. Thx Etienne for this precision.

2017-06-13 6:56 GMT-07:00 Etienne Caron <notifications@github.com>:

**Comment by stephanenicolas** [comment]
@efung would you like to push a PR demonstrating the way you made it work ? It could enhance our base of samples.. :)

**Comment by stephanenicolas** [comment]
We are testing a new approach for this in #232

**Comment by stephanenicolas** [comment]
As an update on this topic, the test orchestrator allows to get a new app instance for every instrumentation test: https://developer.android.com/training/testing/junit-runner.html#using-android-test-orchestrator

**Comment by stephanenicolas** [comment]
I close this topic as it was addressed in 1.1.x
